File: apps/api/src/models/AgentInstruction.ts
import {
  Model,
  DataTypes,
  Optional,
  BelongsToGetAssociationMixin,
} from "sequelize";
import type { Agent } from "./Agent";
import { sequelize } from "../config/database";

export interface AgentInstructionAttributes {
  id: string;
  agentId: string;
  content: string;
  order: number;
  createdAt?: Date;
  updatedAt?: Date;
  isDeleted?: boolean;
  deletedAt?: Date | null;
}

export type AgentInstructionCreationAttributes = Optional<
  AgentInstructionAttributes,
  "id" | "createdAt" | "updatedAt" | "isDeleted" | "deletedAt"
>;

export class AgentInstruction
  extends Model<AgentInstructionAttributes, AgentInstructionCreationAttributes>
  implements AgentInstructionAttributes
{
  declare id: string;
  declare agentId: string;
  declare content: string;
  declare order: number;
  declare readonly createdAt: Date;
  declare readonly updatedAt: Date;
  declare isDeleted: boolean;
  declare deletedAt: Date | null;

  declare getAgent: BelongsToGetAssociationMixin<Agent>;
}

AgentInstruction.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    agentId: {
      type: DataTypes.UUID,
      allowNull: false,
      field: "agent_id",
    },
    content: {
      type: DataTypes.TEXT,
      allowNull: false,
    },
    order: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
    isDeleted: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
      field: "is_deleted",
    },
    deletedAt: {
      type: DataTypes.DATE,
      allowNull: true,
      field: "deleted_at",
    },
  },
  {
    sequelize,
    tableName: "agent_instructions",
    indexes: [{ fields: ["agent_id"] }],
  },
);

File: python/app/models/agent_instruction.py
"""AgentInstruction model: ordered instructions per agent."""

from __future__ import annotations

import uuid

from python.app.models import Agent
from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func

from app.models.base import Base


class AgentInstruction(Base):
    __tablename__ = "agent_instructions"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    agent_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("agents.id", ondelete="CASCADE"), nullable=False, index=True
    )
    content: Mapped[str] = mapped_column(Text(), nullable=False)
    order: Mapped[int] = mapped_column(Integer(), nullable=False)
    is_deleted: Mapped[bool] = mapped_column(Boolean(), server_default="false", nullable=False)
    created_at: Mapped[DateTime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[DateTime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )
    deleted_at: Mapped[DateTime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    agent: Mapped[Agent] = relationship("Agent", back_populates="instructions")
